docs/_docs/02_testing-best-practices/unit-integration-end-to-end-test.md

```diff
@@ -22,3 +22,149 @@ Link to the video at [infoq.com](https://www.infoq.com/presentations/automated-t
+
+If you skip the init_apply step then UnitTest will automatically load your previously applied outputs from the working directory.
+
+The full list of stages is:
+
+* setup
+* init_apply
+* validate
+* teardown
+* destroy
+
+Signature of UnitTest:
+
+```go
+// The output of the module after it has been applied, as returned by terraform.OutputAll
+type ModuleOutputs map[string]interface{}
+
+// A validation function to test your terraform code
+type Validator func(t *testing.T, moduleOutputs *terraform.ModuleOutputs, workingDir string)
+
+type UnitTestPlan struct {
+  TerraformOptions *terraform.Options
+  // If you do not provide a working directory it will default to ./.terratest-unit-test/TEST_FUNCTION_NAME/
+  WorkingDirectory string
+  Setup func(t *testing.T, workingDir string)
+  TearDown func(t *testing.T, moduleOutputs *terraform.ModuleOutputs, workingDir string)
+  Validators []Validator
+}
+
+func UnitTest(t *testing.T, testPlan *terraform.UnitTestPlan)
+```
+
+Examples:
+
+This example shows the simplest example with multiple Validators, with the default working directory
+```go
+func TestWebServer(t *testing.T) {
+  terraformOptions := &terraform.Options {
+    // The path to where your Terraform code is located
+    TerraformDir: "../web-server",
+  }
+
+  terrform.UnitTest(&terraform.UnitTestPlan{
+    TerraformOptions: terraformOptions,
+    Validators: []terraform.Validator{
+      // A list of validation functions to test your module
+      validateWebServer,
+      valudateImageHosting,
+      validateSSLApplied,
+    },
+  })
+}
+
+func validateWebServer(t *testing.T, outputs *terraform.ModuleOutputs, workingDir string) {
+}
+```
+
+This example shows using Setup and Validators, with a custom working directory
+```go
+func TestWebServer(t *testing.T) {
+  terraformOptions := &terraform.Options {
+    // The path to where your Terraform code is located
+    TerraformDir: "../web-server",
+  }
+
+  terrform.UnitTest(&terraform.UnitTestPlan{
+    TerraformOptions: terraformOptions,
+    WorkingDirectory "../web-server",
+    Setup: setup,
+    Validators: []terraform.Validator{
+      // A list of validation functions to test your module
+      validateWebServer,
+      valudateImageHosting,
+      validateSSLApplied,
+    },
+  })
+}
+
+func validateWebServer(t *testing.T, outputs *terraform.ModuleOutputs, workingDir string) {
+  ...
+}
+
+func setup(t *testing.T, workingDir string) {
+  ...
+}
+```
+
+
+## Integration test helpers
+
+**_TODO: As this is a one day project I won't attempt to complete this today _**
+
+Integration tests are very similar to unit tests, only you are testing multiple modules together. To facilitate this
+terraform provides an IntegrationTest function, the main difference to the UnitTest is the IntegrationTestPlan takes a
+list of modules, each with a unique name. It passes a map of ModuleOutputs which are keyed on the unique name you gave
+the module.

```

Comment:
Ah, OK, I see how unit and integration tests compare!

Reply:
What should be the difference of making a integration test of a list with one module? I have the idea that unit tests are for cases where there is no side effect (mostly I/O) and in the case of terraform, we don't do since it's always calling the provider API.

Reply:
@marinalimeira Make sure to watch https://www.infoq.com/presentations/automated-testing-terraform-docker-packer/. With infra code, you're right, all testing is really integration testing... But rather than just throw the term "unit testing" away entirely, it can be useful in this context to think of testing a single module as a "unit test."

Reply:
It's a fine distinction, and in this proposal you could write any unit test as an integration test with a single module, but it requires deeper knowledge and more code, so I think there is real value in providing the more simplified interface of UnitTest since that will cover many cases and allows someone to get started with much less knowledge about the internals of terratest, and maybe less go knowledge too.